op-e2e/actions/l1_replica.go

```diff
@@ -0,0 +1,188 @@
+package actions
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/testutils"
+)
+
+// L1CanonSrc is used to sync L1 from another node.
+// The other node always has the canonical chain.
+// May be nil if there is nothing to sync from
+type L1CanonSrc func(num uint64) *types.Block
+
+// L1Replica is an instrumented in-memory L1 geth node that:
+// - can sync from the given canonical L1 blocks source
+// - can rewind the chain back (for reorgs)
+// - can provide an RPC with mock errors
+type L1Replica struct {
+	log log.Logger
```

Comment:
Does the miner need to be different than the `L1Replica`? It feels like you could make an `L1Node` that has a miner and the `L1Replica` logic

Reply:
Ah nevermind, the miner embeds `L1Node`